x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/ExactMatchProcessor.java

```diff
@@ -57,48 +64,64 @@
-        // If a document does not have the enrich key, return the unchanged document
-        final String value = ingestDocument.getFieldValue(enrichKey, String.class, ignoreMissing);
-        if (value == null) {
-            return ingestDocument;
-        }
-
-        TermQueryBuilder termQuery = new TermQueryBuilder(enrichKey, value);
-        ConstantScoreQueryBuilder constantScore = new ConstantScoreQueryBuilder(termQuery);
-        // TODO: Use a custom transport action instead of the search API
-        SearchSourceBuilder searchBuilder = new SearchSourceBuilder();
-        searchBuilder.size(1);
-        searchBuilder.trackScores(false);
-        searchBuilder.fetchSource(specifications.stream().map(s -> s.sourceField).toArray(String[]::new), null);
-        searchBuilder.query(constantScore);
-
-        SearchRequest req = new SearchRequest();
-        req.indices(EnrichPolicy.getBaseName(policyName));
-        req.preference(Preference.LOCAL.type());
-        req.source(searchBuilder);
-
-        // TODO: Make this Async
-        SearchResponse searchResponse = searchRunner.apply(req);
-
-        // If the index is empty, return the unchanged document
-        // If the enrich key does not exist in the index, throw an error
-        // If no documents match the key, return the unchanged document
-        SearchHit[] searchHits = searchResponse.getHits().getHits();
-        if (searchHits.length < 1) {
-            return ingestDocument;
-        } else if (searchHits.length > 1) {
-            throw new IllegalStateException("more than one doc id matching for [" + enrichKey + "]");
+    public void execute(IngestDocument ingestDocument, BiConsumer<IngestDocument, Exception> handler) {
+        try {
+            // If a document does not have the enrich key, return the unchanged document
+            final String value = ingestDocument.getFieldValue(enrichKey, String.class, ignoreMissing);
+            if (value == null) {
+                handler.accept(ingestDocument, null);
+                return;
+            }
+
+            TermQueryBuilder termQuery = new TermQueryBuilder(enrichKey, value);
+            ConstantScoreQueryBuilder constantScore = new ConstantScoreQueryBuilder(termQuery);
+            // TODO: Use a custom transport action instead of the search API
+            SearchSourceBuilder searchBuilder = new SearchSourceBuilder();
+            searchBuilder.size(1);
+            searchBuilder.trackScores(false);
+            searchBuilder.fetchSource(specifications.stream().map(s -> s.sourceField).toArray(String[]::new), null);
+            searchBuilder.query(constantScore);
+
+            SearchRequest req = new SearchRequest();
+            req.indices(EnrichPolicy.getBaseName(policyName));
+            req.preference(Preference.LOCAL.type());
+            req.source(searchBuilder);
+
+            searchRunner.accept(req, (searchResponse, e) -> {
```

Comment:
If I understand correctly...we are still on the write thread up until here. The search action does some work before it forks and goes async.  Should that work (and later the enrichment work happen) be forked to a different worker (to minimize impact on write and search thread pools) ?

e.g. write -> enrichworker -> search -> enrichworker ?

As it is I think we are doing a bit of search prep work on the write thread and some document enrichment on the search thread. It should be pretty fast work, but not sure if it is ok for it to occupy those slots for that work ?

Reply:
I think as is, is okay. Creating a search request and dealing with a search response is not heavy and there is no need to fork on another thread for those kinds of operations.